### DISCUSSION TITLE
[BE201] Sequelize-CLI  modols 建立後的新增刪除修改

### DISCUSSION THREAD

**Discussion by ddylanlinn** [Lidemy/forum#177] [程式導師實驗計畫] [answered]
大家好，想釐清一下（疑問）自己對於 models 跟 migration 的觀念

建立新 model 流程
1. `$npx sequelize-cli model:generate --name User --attributes username:string,email:string`
（建立 User  table ，columns 包括 username 及 email）

2. 此時會在 models 資料夾建立一個 user model、在 migration 資料夾建立一個 migration 檔案

3. 執行 migration `$npx sequelize-cli db:migrate` 實際建立資料庫欄位

---
若專案開發到一半，想要在table裡新增欄位、修改欄位（型別）、刪除欄位等等更動資料庫結構

假如我今天想在User table 裡新增型別為 STRING 的欄位 `test`

### 方法一，undo
1. 退回先前版本 `$npx sequelize db:migrate:undo `
（版本更動行為依據 migration 檔案裡的 up 跟 down）

2. 編輯 migration 檔案，新增欄位
```js
test: {
 type: Sequelize.STRING
},
```

3. 再跑一次 `$npx sequelize-cli db:migrate`  即可

但是 migration 檔裡預設 down 的內容是 droptable，
所以在執行 `db:migrate:undo` 後整個 table 的資料內容都會丟失

想詢問方法一是正確的嗎？那 down 裡面我要如何修改 undo 後資料內容仍會保留並執行更新？

### 方法二，[建立新的 migration](https://stackoverflow.com/questions/46357533/how-to-add-delete-new-columns-in-sequelize-cli)
1.建立新的 migration 
`$ npx sequelize-cli migration:create --name migration_file_name`

2.編輯新建立 migration 裡的資料
```js
'use strict';
module.exports = {
  up: async (queryInterface, Sequelize) => {
    await queryInterface.addColumn('Users', 'test', {  // 在 Users 下 建立新的欄位 test
      type: Sequelize.STRING, // 型態為 str
      allowNull: true, // 允許空值
    });
  },
  down: async (queryInterface, Sequelize) => {
    await queryInterface.removeColumn('test');  // undo 的話刪掉此欄
  }
};
```

3. 再跑一次 `$npx sequelize-cli db:migrate` ，就可以在資料庫中看到新增的欄位 `test`

---

但是，此時 models 裡的檔案沒有變動
此時我在 controller 裡 user.js 檔案下測試
```js
User.create({
      test: 'this is test', // 測試
      email: 'a@com',
      username: 'name'
    }).then(() => {
      res.send('done')
    })
```
資料庫中新增一筆資料 
username-> name，email-> a@com ，test-> NULL
針對 test 下的 query 沒辦法寫進資料庫（方法一也沒辦法寫入）

於是我在 models 資料夾內的 user model 檔案裡新增欄位
```js
User.init({
    username: DataTypes.STRING,  //原有欄位
    email: DataTypes.STRING,  //原有欄位
    test: DataTypes.STRING //新增此欄
  },
```
再執行一遍相同的測試，此時 `test: 'this is test'` 就順利寫入資料庫欄位中

想詢問，新增 migration 再修改 model 的方法二是對的嗎？（同理修改與刪除也是一樣步驟嗎）

謝謝各位指教🙏

**Comment by cwc329** [comment, marked as answer]
應該要用方法二，
新增、刪除、修改 column, table 也是這樣做。
migration 是用來做資料庫版本控制的工具，
方法二的操作比較符合。

**Comment by aszx87410** [comment]
方法二是對的， migration 只是改 db 而已，model 需要你自己手動調整
